In [1]:
#import modules
import nltk
nltk.download('punkt')
nltk.download('wordnet')
nltk.download('averaged_perceptron_tagger')
nltk.download('stopwords')


import numpy as np


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [2]:
#step2 initi. text
txt="Tokenization is the first step in text analytics. The process of breaking down a text paragraph into smaller chunks such as words or sentences is alled Tokenization."

In [3]:
#Step 3: Perform Tokenization

#Sentence Tokenization

from nltk.tokenize import sent_tokenize

tokenized_text =sent_tokenize(txt)

print (tokenized_text)

#Word Tokenization

from nltk.tokenize import word_tokenize

tokenized_word=word_tokenize(txt)

print (tokenized_word)


['Tokenization is the first step in text analytics.', 'The process of breaking down a text paragraph into smaller chunks such as words or sentences is alled Tokenization.']
['Tokenization', 'is', 'the', 'first', 'step', 'in', 'text', 'analytics', '.', 'The', 'process', 'of', 'breaking', 'down', 'a', 'text', 'paragraph', 'into', 'smaller', 'chunks', 'such', 'as', 'words', 'or', 'sentences', 'is', 'alled', 'Tokenization', '.']


In [6]:
#Step 4: Removing Punctuations and Stop Word

# print stop words of English
from nltk.corpus import stopwords
stop_words=set(stopwords.words("english"))
print(stop_words)

text= "How to remove stop words with NLTK library in Python?"
import re
text= re.sub('[^a-zA-Z]', ' ',text)

tokens = word_tokenize(text.lower())
filtered_text=[]

for w in tokens:
  if w not in stop_words:
    filtered_text.append(w)

print("Tokenized Sentence:", tokens)

print("Filterd Sentence:", filtered_text)

{'what', 'yours', "aren't", 'whom', 'an', 'mightn', "hasn't", 'am', 'ain', 'had', 'nor', 'her', 'each', 'just', 'by', "haven't", 'here', 'as', 'too', 'those', 'me', 'itself', 'after', 'couldn', 's', 'a', 'he', 'against', 'shouldn', 're', 'during', 'until', 'them', 'yourselves', 'such', 'do', 't', "don't", 'where', "she's", 'myself', 'she', 'between', 'aren', 'their', 'was', "couldn't", 'all', 'there', 'few', 'did', "wouldn't", 'through', 'while', 'about', 'over', 'from', 'm', 'you', 'again', 'haven', 'd', 'll', 'when', 'doesn', 'are', 'not', 'hasn', 'how', "you're", 'his', 'any', 'below', 'does', 'so', "won't", 'if', 'yourself', 'but', "doesn't", "isn't", 'didn', 'its', 'this', 'other', "that'll", 'once', 'down', 'up', 'only', "it's", 'some', 'weren', "you'll", 'into', 'ours', 'most', 'y', 'now', "needn't", 'shan', 'hadn', 'or', 'which', 'be', "didn't", "you'd", 'is', 've', 'him', 'under', 'won', 'it', 'then', 'own', "mustn't", 'further', 'more', 'who', 'and', 'why', "mightn't", "shoul

In [8]:
#perform stemming
from nltk.stem import PorterStemmer
e_words= ["wait", "waiting", "waited", "waits"] #example words
ps =PorterStemmer()
for w in e_words:
    rootWord=ps.stem(w)
    print(rootWord)

wait
wait
wait
wait


In [9]:
#perform lemmatization
from nltk.stem import WordNetLemmatizer
wordnet_lemmatizer = WordNetLemmatizer()
text = "studies studying cries cry"
tokenization = nltk.word_tokenize(text)
for w in tokenization:
  print("Lemma for {} is {}".format(w, wordnet_lemmatizer.lemmatize(w)))

Lemma for studies is study
Lemma for studying is studying
Lemma for cries is cry
Lemma for cry is cry


Lemma for studies is study: This line shows that the word "studies" has been correctly lemmatized to its base form "study". The lemmatizer recognized that "studies" is a plural form of the noun "study" and reduced it to its singular form.

Lemma for studying is studying: This line shows that the word "studying" remains unchanged after lemmatization. This is because by default, the WordNetLemmatizer assumes words are nouns. To correctly lemmatize verbs, adjectives, or adverbs, you need to specify the part-of-speech (POS) tag using the pos argument in the lemmatize() method.

Lemma for cries is cry: This line shows that the word "cries" has been correctly lemmatized to its base form "cry". The lemmatizer recognized that "cries" is a verb form and reduced it to its base form.

Lemma for cry is cry: This line shows that the word "cry" remains unchanged after lemmatization. This is because it is already in its base form.

Lemmatization in Text:

In Natural Language Processing (NLP), lemmatization is the process of reducing a word to its base or dictionary form, known as its lemma. This is done by considering the word's context and using morphological analysis to determine its true meaning.

Key points about lemmatization:

Accuracy: Lemmatization is generally more accurate than stemming, which simply chops off prefixes and suffixes.
Context: Lemmatization takes into account the context of the word to determine its lemma.
Morphological Analysis: It uses morphological analysis, which involves examining the structure and form of words, to identify the lemma.
WordNet: The WordNetLemmatizer in NLTK relies on WordNet, a lexical database that provides information about word relationships.
Benefits of lemmatization:

Improved Text Analysis: By reducing words to their base forms, lemmatization helps in improving the accuracy of text analysis tasks such as sentiment analysis, topic modeling, and information retrieval.
Reduced Dimensionality: Lemmatization can reduce the number of unique words in a text, making it easier to process and analyze.
Enhanced NLP Models: Lemmatized text can be used to train more effective NLP models, as it provides a more consistent representation of words.
In essence, lemmatization is a valuable technique for normalizing text data by converting words to their base forms, which ultimately improves the quality and efficiency of NLP tasks. I hope this explanation further clarifies the output and the concept of lemmatization in text. Feel free to ask any more questions you might have.

In [10]:
#apply pos tagging to text
from nltk.tokenize import word_tokenize
data = "The pink sweater fits her perfectly"
words = word_tokenize(data)
for w in words:
  print(nltk.pos_tag([w]))


[('The', 'DT')]
[('pink', 'NN')]
[('sweater', 'NN')]
[('fits', 'NNS')]
[('her', 'PRP$')]
[('perfectly', 'RB')]


gpt expl.
POS Tagging, or Part-of-Speech Tagging, is the process of assigning grammatical tags (also known as POS tags) to each word in a sentence, indicating its part of speech, such as noun, verb, adjective, adverb, etc. It's like labeling each word with its grammatical role in the sentence.

Why is POS tagging important?

Understanding Sentence Structure: POS tags help in understanding the grammatical structure of a sentence, which is crucial for many NLP tasks.
Word Sense Disambiguation: Words can have multiple meanings (e.g., "bank" can refer to a financial institution or a river bank). POS tags can help disambiguate word senses by providing context.
Text Analysis and NLP Tasks: POS tagging is a fundamental step in many NLP applications, including:
Named Entity Recognition (NER): Identifying and classifying named entities like people, organizations, and locations.
Sentiment Analysis: Determining the emotional tone of a text.
Machine Translation: Translating text from one language to another.
Text Summarization: Generating concise summaries of longer texts.
Chatbots and Conversational AI: Building systems that can understand and respond to human language.
How does POS tagging work?

POS tagging involves using algorithms and linguistic rules to analyze the context of a word in a sentence and assign the most likely POS tag. There are different approaches to POS tagging:

Rule-based Tagging: Uses hand-crafted grammatical rules to assign tags.
Statistical Tagging: Employs statistical models trained on large corpora of text to predict the most probable tag based on word frequencies and context.
Deep Learning-based Tagging: Utilizes neural networks to learn complex patterns and relationships between words and their tags.
Example:

Consider the sentence: "The cat sat on the mat."

POS tagging would assign the following tags:

The: Determiner (DET)
cat: Noun (NN)
sat: Verb (VBD)
on: Preposition (IN)
the: Determiner (DET)
mat: Noun (NN)
Libraries for POS Tagging:

NLTK: Provides various POS taggers, including the pos_tag function and the PerceptronTagger.
SpaCy: Offers fast and accurate POS tagging as part of its core NLP pipeline.
Stanford CoreNLP: A comprehensive NLP toolkit with a powerful POS tagger.
In summary, POS tagging is a crucial technique in NLP for understanding the grammatical structure of text and enabling various NLP applications. It provides valuable information about the role of each word in a sentence, allowing for more sophisticated text analysis and processing. I hope this explanation clarifies POS tagging for you. If you have any further questions, feel free to ask!

Python algorithm for creating a TF-IDF (Term Frequency-Inverse Document Frequency) representation of documents.

In [12]:
#Step 1: Import necessary libraries.

from sklearn.feature_extraction.text import TfidfVectorizer

In [13]:
#Step 2: Initialize the documents.
documentA = 'Jupiter is the largest Planet'
documentB = 'Mars is the fourth planet from the Sun'

In [14]:
#Step 3: Create Bag of Words (BoW) for Document A and B.
bagOfWordsA = documentA.split(' ')
bagOfWordsB = documentB.split(' ')
#Step 4: Create a collection of unique words from Document A and B.
uniqueWords = set(bagOfWordsA).union(set(bagOfWordsB))

In [15]:
#Step 5: Create a dictionary of words and their occurrences for each document.
numOfWordsA = dict.fromkeys(uniqueWords, 0)
for word in bagOfWordsA:
    numOfWordsA[word] += 1

numOfWordsB = dict.fromkeys(uniqueWords, 0)
for word in bagOfWordsB:
    numOfWordsB[word] += 1

In [16]:
#Step 6: Compute Term Frequency (TF) for each document
def computeTF(wordDict, bagOfWords):
    tfDict = {}
    bagOfWordsCount = len(bagOfWords)
    for word, count in wordDict.items():
        tfDict[word] = count / float(bagOfWordsCount)
    return tfDict

tfA = computeTF(numOfWordsA, bagOfWordsA)
tfB = computeTF(numOfWordsB, bagOfWordsB)
#Step7 :Compute IDF
def computeIDF(documents):
    import math
    N = len(documents)

    idfDict = dict.fromkeys(documents[0].keys(), 0)
    for document in documents:
        for word, val in document.items():
            if val > 0:
                idfDict[word] += 1

    for word, val in idfDict.items():
        idfDict[word] = math.log(N / float(val))
    return idfDict

idfs = computeIDF([numOfWordsA, numOfWordsB])


In [18]:
#Step 8: Compute TF-IDF for all words
#The computeTFIDF function multiplies each term's TF by its IDF.

def computeTFIDF(tfBagOfWords, idfs):
    tfidf = {}
    for word, val in tfBagOfWords.items():
        tfidf[word] = val * idfs[word]
    return tfidf

tfidfA = computeTFIDF(tfA, idfs)
tfidfB = computeTFIDF(tfB, idfs)
import pandas as pd
df = pd.DataFrame([tfidfA, tfidfB])
df

fourth       Sun   largest   is      from   Jupiter    Planet      Mars  \
0  0.000000  0.000000  0.138629  0.0  0.000000  0.138629  0.138629  0.000000   
1  0.086643  0.086643  0.000000  0.0  0.086643  0.000000  0.000000  0.086643   

   the    planet  
0  0.0  0.000000  
1  0.0  0.086643

Analysis of the Output:
Document 0:

Words like "largest," "Jupiter," and "Planet" have higher TF-IDF scores (0.138629), indicating they are important within this document but do not appear in Document 1.
Words that are absent in this document, such as "fourth," "Sun," and "Mars," have scores of 0.
Document 1:

Words "fourth," "Sun," "Mars," and "planet" have non-zero TF-IDF scores (0.086643), showing their importance in Document 1. These words do not appear in Document 0.
Words like "largest," "Jupiter," and "Planet" have TF-IDF scores of 0, as they are not present in Document 1.